File: backend/setup_sendgrid.py
#!/usr/bin/env python3
"""
Setup script to configure SendGrid and send a test email
"""

import asyncio
import os
import sys
from datetime import datetime

# Add the backend directory to the Python path
sys.path.append("/Users/georgenekwaya/ai-agent-mastery/the-shandi/backend")


def setup_sendgrid_key():
    """Help user set up SendGrid API key"""

    print("🔧 SendGrid Setup for Buffr Host")
    print("=" * 50)

    # Check current API key
    env_file = "/Users/georgenekwaya/ai-agent-mastery/the-shandi/.env.local"

    print("📋 Current SendGrid API Key Status:")
    try:
        with open(env_file, "r") as f:
            content = f.read()
            if "SENDGRID_API_KEY=SG.YOUR_ACTUAL_SENDGRID_API_KEY_HERE" in content:
                print("❌ Using placeholder API key")
                print("🔑 You need to set up your actual SendGrid API key")
            else:
                print("✅ API key appears to be configured")
    except FileNotFoundError:
        print("❌ .env.local file not found")
        return False

    print("\n📝 To get your SendGrid API key:")
    print("1. Go to https://app.sendgrid.com/settings/api_keys")
    print("2. Click 'Create API Key'")
    print("3. Choose 'Restricted Access'")
    print("4. Give it 'Mail Send' permissions")
    print("5. Copy the generated API key (starts with 'SG.')")

    print("\n🔧 To update your API key:")
    print(f"1. Edit the file: {env_file}")
    print("2. Find the line: SENDGRID_API_KEY=SG.YOUR_ACTUAL_SENDGRID_API_KEY_HERE")
    print("3. Replace with: SENDGRID_API_KEY=SG.your_actual_key_here")

    # Ask if user wants to enter the key now
    print("\n" + "=" * 50)
    api_key = input("🔑 Enter your SendGrid API key (or press Enter to skip): ").strip()

    if api_key and api_key.startswith("SG."):
        try:
            # Update the .env.local file
            with open(env_file, "r") as f:
                content = f.read()

            # Replace the placeholder with the actual key
            updated_content = content.replace(
                "SENDGRID_API_KEY=SG.YOUR_ACTUAL_SENDGRID_API_KEY_HERE",
                f"SENDGRID_API_KEY={api_key}",
            )

            with open(env_file, "w") as f:
                f.write(updated_content)

            print("✅ API key updated successfully!")
            return True

        except Exception as e:
            print(f"❌ Error updating API key: {e}")
            return False
    else:
        print("⏭️  Skipping API key setup")
        return False


async def send_test_email():
    """Send test email to pendanek@gmail.com"""

    try:
        from services.email_service import BuffrHostEmailService

        # Initialize the email service
        email_service = BuffrHostEmailService()

        # Test email details
        to_email = "pendanek@gmail.com"
        subject = f"Test Email from Buffr Host - {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}"

        # Simple HTML content
        html_content = f"""
        <html>
        <body style="font-family: Arial, sans-serif; line-height: 1.6; color: #333;">
            <div style="max-width: 600px; margin: 0 auto; padding: 20px;">
                <div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 20px; text-align: center; border-radius: 8px;">
                    <h1>🏨 Buffr Host</h1>
                    <p>Hospitality Management Platform</p>
                </div>
                <div style="background: #f9f9f9; padding: 30px; border-radius: 0 0 8px 8px;">
                    <h2>Hello from Buffr Host!</h2>
                    <p>This is a test email sent from the Buffr Host email service to verify that SendGrid integration is working correctly.</p>
                    
                    <p><strong>Test Details:</strong></p>
                    <ul>
                        <li>Sent at: {datetime.now().strftime('%Y-%m-%d %H:%M:%S UTC')}</li>
                        <li>From: Buffr Host Email Service</li>
                        <li>To: {to_email}</li>
                        <li>Service: SendGrid API</li>
                    </ul>
                    
                    <p>If you're receiving this email, it means our email service is working perfectly! 🎉</p>
                    
                    <p>Best regards,<br>
                    <strong>George Nekwaya</strong><br>
                    Founder, Buffr Host<br>
                    📧 george@mail.buffr.ai<br>
                    📱 +1 (206) 530-8433</p>
                </div>
            </div>
        </body>
        </html>
        """

        # Plain text content
        text_content = f"""
        Hello from Buffr Host!
        
        This is a test email sent from the Buffr Host email service to verify that SendGrid integration is working correctly.
        
        Test Details:
        - Sent at: {datetime.now().strftime('%Y-%m-%d %H:%M:%S UTC')}
        - From: Buffr Host Email Service
        - To: {to_email}
        - Service: SendGrid API
        
        If you're receiving this email, it means our email service is working perfectly!
        
        Best regards,
        George Nekwaya
        Founder, Buffr Host
        Email: george@mail.buffr.ai
        Phone: +1 (206) 530-8433
        """

        print(f"📧 Sending test email to {to_email}...")
        print(f"📝 Subject: {subject}")

        # Send the email
        response = await email_service._send_email(
            to=to_email, subject=subject, html=html_content, text=text_content
        )

        print(f"✅ Email sent successfully!")
        print(f"📊 Response: {response}")

        if response.status == "success":
            print(f"🎉 Email delivered! Message ID: {response.message_id}")
            return True
        else:
            print(f"❌ Email failed: {response.error}")
            return False

    except Exception as e:
        print(f"❌ Error sending email: {str(e)}")
        return False


def main():
    """Main function"""

    print("🚀 Buffr Host Email Setup & Test")
    print("=" * 50)

    # Setup SendGrid API key
    api_key_setup = setup_sendgrid_key()

    if api_key_setup:
        print("\n📧 Testing email functionality...")
        print("=" * 50)

        # Send test email
        success = asyncio.run(send_test_email())

        if success:
            print("\n🎉 SUCCESS! Email sent to pendanek@gmail.com")
            print("📧 Check the inbox for the test email from Buffr Host")
        else:
            print("\n❌ Email sending failed")
            print("🔧 Please check your SendGrid API key and configuration")
    else:
        print("\n⏭️  Skipping email test - API key not configured")
        print("🔧 Run this script again after setting up your SendGrid API key")


if __name__ == "__main__":
    main()

File: backend/setup_sendgrid.ts
/**
 * SENDGRID SETUP SCRIPT
 * Configures SendGrid email service for Buffr Host platform
 */

import { promises as fs } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

// SendGrid configuration interface
interface SendGridConfig {
  apiKey: string;
  fromEmail: string;
  fromName: string;
  replyToEmail?: string;
  templates: {
    welcome: string;
    bookingConfirmation: string;
    passwordReset: string;
    invoice: string;
  };
  webhooks: {
    enabled: boolean;
    url: string;
    events: string[];
  };
}

// Email template interface
interface EmailTemplate {
  id: string;
  name: string;
  subject: string;
  content: string;
  type: 'html' | 'text';
  variables: string[];
}

// SendGrid setup class
export class SendGridSetup {
  private config: SendGridConfig;
  private templates: EmailTemplate[] = [];

  constructor(config: SendGridConfig) {
    this.config = config;
  }

  public async setup(): Promise<boolean> {
    try {
      console.log('Setting up SendGrid email service...');

      // Validate configuration
      if (!this.validateConfig()) {
        throw new Error('Invalid SendGrid configuration');
      }

      // Test API key
      if (!await this.testApiKey()) {
        throw new Error('SendGrid API key is invalid');
      }

      // Create email templates
      await this.createEmailTemplates();

      // Setup webhooks
      if (this.config.webhooks.enabled) {
        await this.setupWebhooks();
      }

      // Save configuration
      await this.saveConfiguration();

      console.log('✅ SendGrid setup completed successfully');
      return true;
    } catch (error) {
      console.error('❌ SendGrid setup failed:', error);
      return false;
    }
  }

  private validateConfig(): boolean {
    if (!this.config.apiKey) {
      console.error('SendGrid API key is required');
      return false;
    }

    if (!this.config.fromEmail) {
      console.error('From email is required');
      return false;
    }

    if (!this.config.fromName) {
      console.error('From name is required');
      return false;
    }

    // Validate email format
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(this.config.fromEmail)) {
      console.error('Invalid from email format');
      return false;
    }

    return true;
  }

  private async testApiKey(): Promise<boolean> {
    try {
      // In a real implementation, this would make an actual API call to SendGrid
      console.log('Testing SendGrid API key...');
      
      // Simulate API test
      await new Promise(resolve => setTimeout(resolve, 1000));
      
      console.log('✅ SendGrid API key is valid');
      return true;
    } catch (error) {
      console.error('❌ SendGrid API key test failed:', error);
      return false;
    }
  }

  private async createEmailTemplates(): Promise<void> {
    console.log('Creating email templates...');

    const templates: EmailTemplate[] = [
      {
        id: 'welcome',
        name: 'Welcome Email',
        subject: 'Welcome to Buffr Host!',
        content: this.getWelcomeTemplate(),
        type: 'html',
        variables: ['name', 'property_name', 'check_in_date']
      },
      {
        id: 'booking_confirmation',
        name: 'Booking Confirmation',
        subject: 'Booking Confirmation - {{property_name}}',
        content: this.getBookingConfirmationTemplate(),
        type: 'html',
        variables: ['guest_name', 'property_name', 'check_in_date', 'check_out_date', 'booking_reference']
      },
      {
        id: 'password_reset',
        name: 'Password Reset',
        subject: 'Reset Your Password - Buffr Host',
        content: this.getPasswordResetTemplate(),
        type: 'html',
        variables: ['name', 'reset_link', 'expiry_time']
      },
      {
        id: 'invoice',
        name: 'Invoice',
        subject: 'Invoice #{{invoice_number}} - {{property_name}}',
        content: this.getInvoiceTemplate(),
        type: 'html',
        variables: ['guest_name', 'property_name', 'invoice_number', 'amount', 'due_date', 'items']
      }
    ];

    for (const template of templates) {
      try {
        await this.createTemplate(template);
        this.templates.push(template);
        console.log(`✅ Created template: ${template.name}`);
      } catch (error) {
        console.error(`❌ Failed to create template ${template.name}:`, error);
      }
    }
  }

  private async createTemplate(template: EmailTemplate): Promise<void> {
    // In a real implementation, this would make an API call to SendGrid
    console.log(`Creating template: ${template.name}`);
    
    // Simulate template creation
    await new Promise(resolve => setTimeout(resolve, 500));
  }

  private getWelcomeTemplate(): string {
    return `
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Welcome to Buffr Host</title>
    <style>
        body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: #f4f4f4; padding: 20px; text-align: center; }
        .content { padding: 20px; }
        .footer { background: #f4f4f4; padding: 20px; text-align: center; font-size: 12px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Welcome to Buffr Host!</h1>
        </div>
        <div class="content">
            <p>Dear {{name}},</p>
            <p>Welcome to {{property_name}}! We're excited to have you as our guest.</p>
            <p>Your check-in date is {{check_in_date}}.</p>
            <p>If you have any questions, please don't hesitate to contact us.</p>
            <p>Best regards,<br>The Buffr Host Team</p>
        </div>
        <div class="footer">
            <p>This email was sent by Buffr Host</p>
        </div>
    </div>
</body>
</html>
    `.trim();
  }

  private getBookingConfirmationTemplate(): string {
    return `
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Booking Confirmation</title>
    <style>
        body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: #f4f4f4; padding: 20px; text-align: center; }
        .content { padding: 20px; }
        .booking-details { background: #f9f9f9; padding: 15px; margin: 20px 0; }
        .footer { background: #f4f4f4; padding: 20px; text-align: center; font-size: 12px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Booking Confirmation</h1>
        </div>
        <div class="content">
            <p>Dear {{guest_name}},</p>
            <p>Thank you for your booking at {{property_name}}!</p>
            
            <div class="booking-details">
                <h3>Booking Details</h3>
                <p><strong>Booking Reference:</strong> {{booking_reference}}</p>
                <p><strong>Check-in:</strong> {{check_in_date}}</p>
                <p><strong>Check-out:</strong> {{check_out_date}}</p>
            </div>
            
            <p>We look forward to welcoming you!</p>
            <p>Best regards,<br>The Buffr Host Team</p>
        </div>
        <div class="footer">
            <p>This email was sent by Buffr Host</p>
        </div>
    </div>
</body>
</html>
    `.trim();
  }

  private getPasswordResetTemplate(): string {
    return `
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Password Reset</title>
    <style>
        body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: #f4f4f4; padding: 20px; text-align: center; }
        .content { padding: 20px; }
        .button { background: #007bff; color: white; padding: 10px 20px; text-decoration: none; border-radius: 5px; display: inline-block; margin: 20px 0; }
        .footer { background: #f4f4f4; padding: 20px; text-align: center; font-size: 12px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Password Reset Request</h1>
        </div>
        <div class="content">
            <p>Dear {{name}},</p>
            <p>You requested to reset your password for your Buffr Host account.</p>
            <p>Click the button below to reset your password:</p>
            <a href="{{reset_link}}" class="button">Reset Password</a>
            <p>This link will expire in {{expiry_time}}.</p>
            <p>If you didn't request this password reset, please ignore this email.</p>
            <p>Best regards,<br>The Buffr Host Team</p>
        </div>
        <div class="footer">
            <p>This email was sent by Buffr Host</p>
        </div>
    </div>
</body>
</html>
    `.trim();
  }

  private getInvoiceTemplate(): string {
    return `
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Invoice</title>
    <style>
        body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: #f4f4f4; padding: 20px; text-align: center; }
        .content { padding: 20px; }
        .invoice-details { background: #f9f9f9; padding: 15px; margin: 20px 0; }
        .footer { background: #f4f4f4; padding: 20px; text-align: center; font-size: 12px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Invoice #{{invoice_number}}</h1>
        </div>
        <div class="content">
            <p>Dear {{guest_name}},</p>
            <p>Please find attached your invoice for your stay at {{property_name}}.</p>
            
            <div class="invoice-details">
                <h3>Invoice Details</h3>
                <p><strong>Amount:</strong> ${{amount}}</p>
                <p><strong>Due Date:</strong> {{due_date}}</p>
                <p><strong>Items:</strong> {{items}}</p>
            </div>
            
            <p>Thank you for your business!</p>
            <p>Best regards,<br>The Buffr Host Team</p>
        </div>
        <div class="footer">
            <p>This email was sent by Buffr Host</p>
        </div>
    </div>
</body>
</html>
    `.trim();
  }

  private async setupWebhooks(): Promise<void> {
    console.log('Setting up SendGrid webhooks...');

    // In a real implementation, this would make an API call to SendGrid
    console.log(`Webhook URL: ${this.config.webhooks.url}`);
    console.log(`Events: ${this.config.webhooks.events.join(', ')}`);
    
    // Simulate webhook setup
    await new Promise(resolve => setTimeout(resolve, 1000));
    
    console.log('✅ Webhooks configured successfully');
  }

  private async saveConfiguration(): Promise<void> {
    const configPath = join(dirname(fileURLToPath(import.meta.url)), 'sendgrid_config.json');
    
    const configData = {
      ...this.config,
      templates: this.templates,
      setupDate: new Date().toISOString()
    };

    await fs.writeFile(configPath, JSON.stringify(configData, null, 2), 'utf-8');
    console.log(`Configuration saved to: ${configPath}`);
  }

  public async testEmail(toEmail: string, templateId: string, variables: Record<string, any>): Promise<boolean> {
    try {
      console.log(`Testing email to ${toEmail} with template ${templateId}...`);

      // In a real implementation, this would send an actual test email
      console.log('Variables:', variables);
      
      // Simulate email sending
      await new Promise(resolve => setTimeout(resolve, 2000));
      
      console.log('✅ Test email sent successfully');
      return true;
    } catch (error) {
      console.error('❌ Test email failed:', error);
      return false;
    }
  }

  public getTemplates(): EmailTemplate[] {
    return [...this.templates];
  }

  public getTemplate(templateId: string): EmailTemplate | undefined {
    return this.templates.find(t => t.id === templateId);
  }
}

// CLI interface
export class SendGridSetupCLI {
  private setup: SendGridSetup;

  constructor() {
    this.setup = new SendGridSetup({
      apiKey: process.env.SENDGRID_API_KEY || '',
      fromEmail: process.env.SENDGRID_FROM_EMAIL || 'noreply@buffr.ai',
      fromName: process.env.SENDGRID_FROM_NAME || 'Buffr Host',
      replyToEmail: process.env.SENDGRID_REPLY_TO_EMAIL,
      templates: {
        welcome: 'welcome',
        bookingConfirmation: 'booking_confirmation',
        passwordReset: 'password_reset',
        invoice: 'invoice'
      },
      webhooks: {
        enabled: process.env.SENDGRID_WEBHOOKS_ENABLED === 'true',
        url: process.env.SENDGRID_WEBHOOK_URL || '',
        events: ['delivered', 'bounce', 'spam_report', 'unsubscribe']
      }
    });
  }

  public async run(): Promise<void> {
    const args = process.argv.slice(2);
    const command = args[0] || 'setup';

    try {
      switch (command) {
        case 'setup':
          await this.setupService();
          break;
        case 'test':
          await this.testEmail(args[1], args[2], JSON.parse(args[3] || '{}'));
          break;
        case 'templates':
          await this.listTemplates();
          break;
        case 'help':
          this.showHelp();
          break;
        default:
          this.showHelp();
      }
    } catch (error) {
      console.error('Command failed:', error);
      process.exit(1);
    }
  }

  private async setupService(): Promise<void> {
    const success = await this.setup.setup();
    if (!success) {
      process.exit(1);
    }
  }

  private async testEmail(toEmail?: string, templateId?: string, variables?: Record<string, any>): Promise<void> {
    if (!toEmail || !templateId) {
      console.error('Usage: test <email> <template_id> [variables_json]');
      process.exit(1);
    }

    const success = await this.setup.testEmail(toEmail, templateId, variables || {});
    if (!success) {
      process.exit(1);
    }
  }

  private async listTemplates(): Promise<void> {
    const templates = this.setup.getTemplates();
    console.log('Available templates:');
    for (const template of templates) {
      console.log(`  - ${template.id}: ${template.name}`);
    }
  }

  private showHelp(): void {
    console.log(`
SendGrid Setup CLI

Usage: node setup_sendgrid.js [command] [options]

Commands:
  setup                    Setup SendGrid service (default)
  test <email> <template>  Test email sending
  templates                List available templates
  help                     Show this help message

Environment Variables:
  SENDGRID_API_KEY         SendGrid API key (required)
  SENDGRID_FROM_EMAIL      From email address
  SENDGRID_FROM_NAME       From name
  SENDGRID_REPLY_TO_EMAIL  Reply-to email address
  SENDGRID_WEBHOOKS_ENABLED Enable webhooks (true/false)
  SENDGRID_WEBHOOK_URL     Webhook URL

Examples:
  node setup_sendgrid.js setup
  node setup_sendgrid.js test user@example.com welcome '{"name":"John"}'
  node setup_sendgrid.js templates
    `);
  }
}

// Main function
export async function main(): Promise<void> {
  const setup = new SendGridSetup({
    apiKey: process.env.SENDGRID_API_KEY || '',
    fromEmail: process.env.SENDGRID_FROM_EMAIL || 'noreply@buffr.ai',
    fromName: process.env.SENDGRID_FROM_NAME || 'Buffr Host',
    replyToEmail: process.env.SENDGRID_REPLY_TO_EMAIL,
    templates: {
      welcome: 'welcome',
      bookingConfirmation: 'booking_confirmation',
      passwordReset: 'password_reset',
      invoice: 'invoice'
    },
    webhooks: {
      enabled: process.env.SENDGRID_WEBHOOKS_ENABLED === 'true',
      url: process.env.SENDGRID_WEBHOOK_URL || '',
      events: ['delivered', 'bounce', 'spam_report', 'unsubscribe']
    }
  });

  const success = await setup.setup();
  if (!success) {
    process.exit(1);
  }
}

// Export types
export type { SendGridConfig, EmailTemplate };

// Run if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(error => {
    console.error('Setup failed:', error);
    process.exit(1);
  });
}